Shared/sdk/SharedUtil.Template.h

```diff
@@ -0,0 +1,193 @@
+    using param2_t = Arg2;
+};
+
+template <typename Test>
+struct is_variant : std::false_type
+{
+};
+
+template <typename Arg1, typename... Args>
+struct is_variant<std::variant<Arg1, Args...>> : std::true_type
+{
+    using param1_t = Arg1;
+    using rest_t = std::variant<Args...>;
+    static constexpr auto count = sizeof...(Args) + 1;
+};
+
+/**
+    nth_element
+
+    Returns the nth element of a parameter pack
+**/
+
+template <std::size_t I, typename T, typename... Ts>
+struct nth_element_impl
+{
+    using type = typename nth_element_impl<I - 1, Ts...>::type;
+};
+
+template <typename T, typename... Ts>
+struct nth_element_impl<0, T, Ts...>
+{
+    using type = T;
+};
+
+template <std::size_t I, typename... Ts>
+using nth_element_t = typename nth_element_impl<I, Ts...>::type;
+
+
+// common_variant
+// common_variant builds a single variant from types or variants
+//  bool, bool -> variant<bool>
+//  int, bool -> variant<int, bool>
+//  variant<int, bool>, bool -> variant<int, bool>
+//  variant<int, bool>, float -> variant<int, bool, float>
+//  variant<int, bool>, variant<bool, float> -> variant<int, bool, float>
+//  variant<int, bool>, variant<bool, float, double> -> variant<int, bool, float, double>
+
+// Two different types
+//  int, bool -> variant<int, bool>
+template <typename T, typename U>
+struct common_variant
+{
+    using type = std::variant<T, U>;
+};
+
+// Identical types
+//  int, int -> variant<int>
+template <typename T>
+struct common_variant<T, T>
+{
+    using type = std::variant<T>;
+};
+
+// Type + Variant which starts with the Type
+//  int, variant<int> -> variant<int>
+template <typename T, typename... Ts>
+struct common_variant<T, std::variant<T, Ts...>>
+{
+    using type = std::variant<T, Ts...>;
+};
+
+// Variant + Empty = Variant
+template <typename... Ts>
+struct common_variant<std::variant<Ts...>, std::variant<>>
+{
+    using type = std::variant<Ts...>;
+};
+// Empty + Variant = Variant
+template <typename... Ts>
+struct common_variant<std::variant<>, std::variant<Ts...>>
+{
+    using type = std::variant<Ts...>;
+};
+
+template <typename T, typename... Ts>
+struct common_variant<T, std::variant<Ts...>>
+{
+    using type = std::conditional_t<std::is_convertible_v<T, std::variant<Ts...>>, std::variant<Ts...>, std::variant<T, Ts...>>;
+};
+
+template <typename T, typename... Ts>
+struct common_variant<std::variant<Ts...>, T>
+{
+    using type = typename common_variant<T, std::variant<Ts...>>::type;
+};
+
+template <typename T, typename... Ts, typename... Us>
+struct common_variant<std::variant<T, Ts...>, std::variant<Us...>>
+{
+    using type = typename common_variant<std::variant<Ts...>, typename common_variant<T, std::variant<Us...>>::type>::type;
+};
+
+// dummy_type
+// generic dummy type
+struct dummy_type
+{
+};
+
+
+// n_tuple: Constructs a tuple of size N
+//  n_tuple<2>::type == std::tuple<dummy_type, dummy_type>
+template <std::size_t, bool B = false, typename... Args>
```

Comment:
What does `B` mean?

At a deeper look I think it might mean `Base case`. If that's correct, consider making this use of recursion obvious.

```suggestion
template <std::size_t, bool BaseCase = false, typename... Args>
```

Reply:
Yeah, it's somewhat of a base case. Basically it is used to figure out whether more Arguments must be added in the next step.